src/tools/rbd_wnbd/wnbd_wmi.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (c) 2019 SUSE LLC
+ * Copyright (C) 2022 Cloudbase Solutions
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+#include <comutil.h>
+
+#define _WIN32_DCOM
+#include <wbemcli.h>
+
+#include <string>
+#include <vector>
+
+HRESULT co_initialize_basic();
+
+class WmiConnection
+{
+private:
+  std::wstring ns;
+public:
+  IWbemLocator* wbem_loc;
+  IWbemServices* wbem_svc;
+
+  WmiConnection()
+    : wbem_loc(nullptr)
+    , wbem_svc(nullptr)
+  {
+  }
+  WmiConnection(std::wstring ns)
+    : ns(ns)
+    , wbem_loc(nullptr)
+    , wbem_svc(nullptr)
+  {
+  }
+  ~WmiConnection()
+  {
+    close();
+  }
+
+  HRESULT initialize();
+  void close();
+};
+
+HRESULT get_property_str(
+  IWbemClassObject* cls_obj,
+  const std::wstring& property,
+  std::wstring& value);
+HRESULT get_property_int(
+  IWbemClassObject* cls_obj,
+  const std::wstring& property,
+  uint32_t& value);
+
+class WmiSubscription
+{
+private:
+  std::wstring ns;
+  std::wstring query;
+
+  WmiConnection conn;
+  IEnumWbemClassObject *event_enum;
+
+public:
+  WmiSubscription(std::wstring ns, std::wstring query)
+    : ns(ns)
```

Comment:
Does `WmiSubscription` actually need `ns` member?  Could this argument just be passed `WmiConnection` constructor?

Reply:
Right, initially I thought we might log it but it can be dropped.